### PR TITLE
Fix error routes

### PR DESCRIPTION
**Pull request by nkzawa**
- Fix 404 is not working client side
- Possibly fix hot code reload has issues that break SSR

### CODE REVIEW

File: server/render.js [hunk 1]
```diff
@@ -168,15 +168,7 @@ export async function serveStaticWithGzip (req, res, path) {
...
-    .on('error', (err) => {
-      if (err.code === 'ENOENT') {
-        res.statusCode = 404
-        res.end('Not Found')
-        resolve()
-      } else {
-        reject(err)
-      }
-    })
+    .on('error', reject)
```
Comment (nkzawa) [inline server/render.js]: We have to respond with JSON of error component if it's a JSON route.

File: lib/router/router.js [hunk 2]
```diff
@@ -127,16 +135,21 @@ export default class Router extends EventEmitter {
       data, props, error
     } = await this.getRouteInfo(route, pathname, query)
 
-    if (error) {
+    if (error && error.cancelled) {
       this.emit('routeChangeError', error, as)
-      throw error
+      return false
     }
```
Comment (arunoda) [inline lib/router/router.js]: With this, I think it make no sense to emit `routeChangeError`.
And we could simply avoid sending all abort events.

File: server/render.js (hunk 1 above)
Comment (arunoda) [inline server/render.js]: How about changing `renderJSON` of the `index.js` into this:

```js
async renderJSON (req, res, page) {
    if (this.dev) {
      const compilationErr = this.getCompilationError(page)
      if (compilationErr) {
        return this.renderErrorJSON(compilationErr, req, res)
      }
    }

    try {
      await renderJSON(req, res, page, this.renderOpts)
    } catch (err) {
      if (err.code === 'ENOENT') {
        res.statusCode = 404
        res.end('404')
      } else {
        if (!this.quiet) console.error(err)
        res.statusCode = 500
        res.end('500')
      }
    }
  }
```

File: lib/router/router.js (hunk 2 above)
Comment (nkzawa) [inline lib/router/router.js]: You mean we notify nothing if it's aborted ? Anyway it should be another issue, I don't want to change the behavior on this PR.

File: server/render.js (hunk 1 above)
Comment (nkzawa) [inline server/render.js]: We have to return a JSON, not a text. That's the reason of this fix.

**Pull request merged**